src/rowbinary/de.rs

```diff
@@ -1,30 +1,75 @@
-use std::{convert::TryFrom, mem, str};
-
 use crate::error::{Error, Result};
+use crate::row_metadata::RowMetadata;
+use crate::rowbinary::utils::{ensure_size, get_unsigned_leb128};
+use crate::rowbinary::validation::SerdeType;
+use crate::rowbinary::validation::{DataTypeValidator, SchemaValidator};
+use crate::Row;
 use bytes::Buf;
+use core::mem::size_of;
+use serde::de::MapAccess;
 use serde::{
     de::{DeserializeSeed, Deserializer, EnumAccess, SeqAccess, VariantAccess, Visitor},
     Deserialize,
 };
+use std::marker::PhantomData;
+use std::{convert::TryFrom, str};
 
-/// Deserializes a value from `input` with a row encoded in `RowBinary`.
+/// Deserializes a value from `input` with a row encoded in `RowBinary`,
+/// i.e. only when [`crate::Row`] validation is disabled in the client.
 ///
 /// It accepts _a reference to_ a byte slice because it somehow leads to a more
 /// performant generated code than `(&[u8]) -> Result<(T, usize)>` and even
 /// `(&[u8], &mut Option<T>) -> Result<usize>`.
-pub(crate) fn deserialize_from<'data, T: Deserialize<'data>>(input: &mut &'data [u8]) -> Result<T> {
-    let mut deserializer = RowBinaryDeserializer { input };
+///
+/// Additionally, having a single function speeds up [`crate::cursors::RowCursor::next`] x2.
+/// A hint about the [`Error::NotEnoughData`] gives another 20% performance boost.
+///
+/// It expects a slice of [`Column`] objects parsed
+/// from the beginning of `RowBinaryWithNamesAndTypes` data stream.
+/// After the header, the rows format is the same as `RowBinary`.
+pub(crate) fn deserialize_row_binary<'data, 'cursor, T: Deserialize<'data> + Row>(
+    input: &mut &'data [u8],
+) -> Result<T> {
+    let mut deserializer = RowBinaryDeserializer::<T, _>::new(input, ());
+    T::deserialize(&mut deserializer)
+}
+
+/// Similar to [`deserialize_row_binary`], but uses [`RowMetadata`]
+/// parsed from `RowBinaryWithNamesAndTypes` header to validate the data types.
+/// This is used when [`crate::Row`] validation is enabled in the client (default).
+pub(crate) fn deserialize_rbwnat<'data, 'cursor, T: Deserialize<'data> + Row>(
```

Comment:
So, is it actually `row_binary`, not `rbwnat` (because columns are decoded already)?